swarm/src/dummy.rs

```diff
@@ -58,7 +58,6 @@ pub struct ConnectionHandler;
 impl crate::handler::ConnectionHandler for ConnectionHandler {
     type InEvent = Void;
     type OutEvent = Void;
-    type Error = Void;
```

Comment:
I find this type information useful. I.e. I believe there is value in making it impossible at compile time for a protocol to close a connection.

Reply:
Can you articulate and perhaps quantify the value you are seeing?

In my view, type systems are best leveraged on boundaries. Boundaries are where we likely split work between developers (users vs maintainers for example) so they have documentation value and the compiler can ensure invariants.

In this case, the `Error` is part of a boundary but the consumer is within our responsibility (`Swarm`). Because it is an associated type, we have to cater for it to be anything so we can't really take advantage of the type system here. On the contrary, it makes things more complicated because we have to carry around a type parameter which is a cost in terms of complexity.

In addition to `Swarm`, this boundary also touches the implementer of a `ConnectionHandler` and the user of `Swarm`.

Let's start with the latter. Connections can be closed at any time for a variety of reasons. There is definitely value in getting notified about connections being closed. However, I'd argue that it is unlikely that one would take different codepaths depending on which part of my system actively closed a connection. A type-safe solution is useful if I want the compiler to ensure how and when certain code-paths are being taken. But, if we treat all events about closed connections the same, there is no added value in a type-system driven solution.

Last but not least, we have the implementer of `ConnectionHandler`. One may argue that it is a useful safety check to first declare that a protocol may never close a connection to ensure this isn't accidentally introduced later (either by the same dev or as part of a change by someone else). This is a plus for a type-system driven solution although it is important to add that one can still audit fairly easily whether a particular protocol actively closes a connection or not.

An additional negative point is that more moving parts (i.e. number of associated types) makes it more difficult for newcomers to implement a `ConnectionHandler`. By how much is difficult to measure but from my own experience, I was quite overwhelmed with this, esp. because the default value for when you generate this trait implementation (`type Error = ()`) doesn't compile.

In my view, this boils down to weighing the ease of auditability of protocols for active connection closings vs the complexity this type parameter causes in our codebase.

To me, reducing the complexity in our codebase is a pretty clear winner here. It is difficult to express this in numbers but I'd assume that auditing a protocol for when it closes connections is something that doesn't happen very often. Compared to that, we have 4 maintainers and various active contributers here that routinely come in contact with the parts of this codebase involving this type parameter.

Reply:
What I forgot to add: A pretty substantial cost is that this type appears within `SwarmEvent` too which pretty much every user of `libp2p` comes in contact with and is clunky to name due to the composition of the error as part of `NetworkBehaviour`.

Reply:
I agree with the reasoning you lay out above. I do believe there is value in strong typing here, e.g. preventing at compile time that a `ConnectionHandler` never closes a connection. That said, I am not sure it warrants the downsides you list above. Long story short, I don't have a strong opinion here. Let's go with your suggestion.